azurerm/resource_arm_mysql_virtual_network_rule_test.go

```diff
@@ -0,0 +1,476 @@
+	})
+}
+
+func TestAccAzureRMMySqlVirtualNetworkRule_switchSubnets(t *testing.T) {
+	resourceName := "azurerm_mysql_virtual_network_rule.test"
+	ri := acctest.RandInt()
+
+	preConfig := testAccAzureRMMySqlVirtualNetworkRule_subnetSwitchPre(ri, testLocation())
+	postConfig := testAccAzureRMMySqlVirtualNetworkRule_subnetSwitchPost(ri, testLocation())
+
+	// Create regex strings that will ensure that one subnet name exists, but not the other
+	preConfigRegex := regexp.MustCompile(fmt.Sprintf("(subnet1%d)$|(subnet[^2]%d)$", ri, ri))  //subnet 1 but not 2
+	postConfigRegex := regexp.MustCompile(fmt.Sprintf("(subnet2%d)$|(subnet[^1]%d)$", ri, ri)) //subnet 2 but not 1
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMMySqlVirtualNetworkRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: preConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName),
+					resource.TestMatchResourceAttr(resourceName, "subnet_id", preConfigRegex),
+				),
+			},
+			{
+				Config: postConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName),
+					resource.TestMatchResourceAttr(resourceName, "subnet_id", postConfigRegex),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAzureRMMySqlVirtualNetworkRule_disappears(t *testing.T) {
+	resourceName := "azurerm_mysql_virtual_network_rule.test"
+	ri := acctest.RandInt()
+	config := testAccAzureRMMySqlVirtualNetworkRule_basic(ri, testLocation())
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMMySqlVirtualNetworkRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName),
+					testCheckAzureRMMySqlVirtualNetworkRuleDisappears(resourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
+func TestAccAzureRMMySqlVirtualNetworkRule_multipleSubnets(t *testing.T) {
+	resourceName1 := "azurerm_mysql_virtual_network_rule.rule1"
+	resourceName2 := "azurerm_mysql_virtual_network_rule.rule2"
+	resourceName3 := "azurerm_mysql_virtual_network_rule.rule3"
+	ri := acctest.RandInt()
+	config := testAccAzureRMMySqlVirtualNetworkRule_multipleSubnets(ri, testLocation())
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testCheckAzureRMMySqlVirtualNetworkRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: config,
+				Check: resource.ComposeTestCheckFunc(
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName1),
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName2),
+					testCheckAzureRMMySqlVirtualNetworkRuleExists(resourceName3),
+				),
+			},
+		},
+	})
+}
+
```

Comment:
Since they allow for in place SKU update should we add a test that:

**Test Steps:**
1. Create the basic MySQL server as in `TestAccAzureRMMySqlVirtualNetworkRule_basic`
2. Updates the SKU and validates that the rules also remain after the update?

This would have to be a `General Purpose` SKU that is upgraded to an `Optimized` SKU, so it the test would not be able to re-use the `TestAccAzureRMMySqlVirtualNetworkRule_basic` configuration file.

Reply:
IMO that check (that the MySQL Server isn't created when the SKU changes) belongs on the Server resource, rather than on this one? In other providers where the ID's are unique we check the ID of the resource before and after making the change to ensure they're the same, but where Azure uses the name field as the ID we'd need to look into using a different field to verify this is the same resource?